### DISCUSSION TITLE
viewCreator2d

### DISCUSSION THREAD

**Discussion by paprophet** [iTwin/itwinjs-core#323] [Q&A] [answered]
viewCreator2d.createViewForModel(modelId: Id64String, modelType: string, options?: ViewCreator2dOptions)
when modelType set "BisCore:SpatialViewDefinition"  the show error message :modelType not supported
what modelType  value  is right？

**Comment by paprophet** [comment, marked as answer]
const viewCreator = new ViewCreator2d(imodel);
const models = await imodel.models.queryProps({ from: "BisCore.GeometricModel2d" });
if (models.length > 0)
  const view = await viewCreator.createViewForModel(models[0].id!, models[0].classFullName);

what format file to bim file  have "BisCore.GeometricModel2d"?

**Reply by pmconne** [reply]
`BisCore.GeometricModel2d` is a 2d model - i.e., a drawing or a sheet. The source file format does not matter - whether coming from Revit or DGN or DWG or anywhere else, 2d models become `GeometricModel2d`s.
Appropriate `modelType`s for 2d models are `DrawingModelState.classFullName`, `SectionDrawingModelState.classFullName`, and `SheetModelState.classFullName`.
The need to specify a "model type" for these APIs seems awkward to me - @roopksaini?

**Reply by roopksaini** [reply]
It's there to distinguish a sheet view from a drawing/section since sheet views require additional props. It might be better to instead add an optional parameter "isSheetView". Another idea is to use the modelId to queryProps and handle type determination within the API. Think originally that seemed redundant since the user would already have that info before calling the API.

 const models = await imodel.models.queryProps({ from: "BisCore.GeometricModel2d" });
    let viewState2d;
    if (models.length > 0)
      viewState2d = await viewCreator.getViewForModel(models[0].id!, models[0].classFullName, { bgColor: ColorDef.black });

But now that I think about it...they could have just the model ids cached. Also, if I check for types within the API, I'd have an opportunity to throw an exception if a non-2d model ID is passed. I think I just convinced myself the second approach is the way to go. Any objections? @pmconne

**Reply by pmconne** [reply]
Querying the model type from the backend is what I was thinking. I suppose you could have an *optional* `modelType` argument for cases where the caller already knows the type. If so, it would be preferable to use an enum or a union of string literals instead of accepting an arbitrary (and verbose) `string` - there are only 3 valid inputs.

**Reply by pmconne** [reply]
My suggestion is to get rid of `modelType` entirely. Keep it simple.
